whylog/teacher/constraint_links_base.py

```diff
@@ -0,0 +1,60 @@
+class ConstraintLinksBase:
+    """
+    Container for links between group and constraint
+    """
+
+    LINE_ID_POSITION = 0
+    GROUP_NUMBER_POSITION = 1
+    CONSTRAINT_ID_POSITION = 2
+
+    def __init__(self, links_list=None):
+        """
+        :param links_list: list of links between group and constraint
+                          one link is a tuple (line_id, group_no, constraint_id)
+        :type links_list: list[(int, int, int)]
+        """
+        self.links = links_list or []
+
+    def get_links(self):
+        return self.links
+
+    def distinct_constraint_ids(self):
+        return set(list(zip(*self.links))[self.CONSTRAINT_ID_POSITION])
+
+    def add_links(self, links):
+        self.links = list(set(self.links).union(links))
+
+    def remove_links_by_line(self, line_id):
+        self._remove_links_where(line_id=line_id)
+
+    def remove_links_by_group(self, line_id, group_no):
+        self._remove_links_where(line_id=line_id, group_no=group_no)
+
+    def remove_links_by_constraint(self, constraint_id):
+        self._remove_links_where(constr_id=constraint_id)
+
+    def _remove_links_where(self, line_id=None, group_no=None, constr_id=None):
```

Comment:
Use `filter()` instead of the `_select` monstrosity. Filtering functions should be defined in scope of module and passed as a parameter by `remove_links_by_*` functions to here.